UI/obs-app-theming.cpp

```diff
@@ -0,0 +1,962 @@
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+#include <cinttypes>
+
+#include <util/cf-parser.h>
+
+#include <QDir>
+#include <QFile>
+#include <QTimer>
+#include <QMetaEnum>
+#include <QDirIterator>
+#include <QGuiApplication>
+#include <QRandomGenerator>
+
+#include "qt-wrappers.hpp"
+#include "obs-app.hpp"
+#include "obs-app-theming.hpp"
+#include "obs-proxy-style.hpp"
+#include "platform.hpp"
+
+#include "ui-config.h"
+
+using namespace std;
+
+struct CFParser {
+	cf_parser cfp = {};
+	~CFParser() { cf_parser_free(&cfp); }
+	operator cf_parser *() { return &cfp; }
+	cf_parser *operator->() { return &cfp; }
+};
+
+static OBSTheme *ParseThemeMeta(const QString &path)
+{
+	QFile themeFile(path);
+	if (!themeFile.open(QIODeviceBase::ReadOnly))
+		return nullptr;
+
+	OBSTheme *meta = nullptr;
+	const QByteArray data = themeFile.readAll();
+	CFParser cfp;
+	int ret;
+
+	if (!cf_parser_parse(cfp, data.constData(), QT_TO_UTF8(path)))
+		return nullptr;
+
+	if (cf_token_is(cfp, "OBSThemeMeta") ||
+	    cf_go_to_token(cfp, "OBSThemeMeta", nullptr)) {
+
+		if (!cf_next_token(cfp))
+			return nullptr;
+
+		if (!cf_token_is(cfp, "{"))
+			return nullptr;
+
+		meta = new OBSTheme();
+
+		for (;;) {
+			if (!cf_next_token(cfp)) {
+				delete meta;
+				return nullptr;
+			}
+
+			ret = cf_token_is_type(cfp, CFTOKEN_NAME, "name",
+					       nullptr);
+			if (ret != PARSE_SUCCESS)
+				break;
+
+			string name(cfp->cur_token->str.array,
+				    cfp->cur_token->str.len);
+
+			ret = cf_next_token_should_be(cfp, ":", ";", nullptr);
+			if (ret != PARSE_SUCCESS)
+				continue;
+
+			if (!cf_next_token(cfp)) {
+				delete meta;
+				return nullptr;
+			}
+
+			ret = cf_token_is_type(cfp, CFTOKEN_STRING, "value",
+					       ";");
+
+			if (ret != PARSE_SUCCESS)
+				continue;
+
+			BPtr str = cf_literal_to_str(cfp->cur_token->str.array,
+						     cfp->cur_token->str.len);
+
+			if (str) {
+				if (name == "dark")
+					meta->isDark = strcmp(str, "true") == 0;
+				else if (name == "extends")
+					meta->extends = str;
+				else if (name == "author")
+					meta->author = str;
+				else if (name == "id")
+					meta->id = str;
+				else if (name == "name")
+					meta->name = str;
+			}
+
+			if (!cf_go_to_token(cfp, ";", nullptr)) {
+				delete meta;
+				return nullptr;
+			}
+		}
+	}
+
+	if (meta) {
+		meta->isBaseTheme = path.endsWith(".obt");
+
+		if (meta->id.isEmpty() || meta->name.isEmpty() ||
+		    (!meta->isBaseTheme && meta->extends.isEmpty())) {
+			/* Theme is invalid */
+			delete meta;
+			meta = nullptr;
+		} else {
+			BPtr absPath = os_get_abs_path_ptr(QT_TO_UTF8(path));
+			meta->location = absPath;
+			meta->isHighContrast = path.endsWith(".oha");
+			meta->isVisible = !path.contains("System");
+		}
+	}
+
+	return meta;
+}
+
+static QColor ParseColour(CFParser &cfp)
```

Comment:
I know somebody who will tell you to use US-American spelling here.. ;)

Reply:
Indeed.